Initial commit. Added support for register confirmation and personal …

grumpykiwi opened the pull request:
Initial support for Register Confirmation and Personal Data download/delete.

grumpykiwi committed:
Initial commit. Added support for register confirmation and personal data

grumpykiwi committed:
Added new account confirmation, account lockout, 2FA (with caveats - refer readme), data download, delete account, return url

DamianEdwards commented on 2FANotes.txt:
```diff
@@ -0,0 +1,55 @@
+
+It is important to mention that as soon as a user gets redirected to the LoginTwoStep action, the new Identity.TwoFactorUserId cookie will be created in our browser. This cookie contains important data about the current user.
+
+I am suspicious this has something to do with what is going on. 
```
Yeah you can't write out cookies from the Blazor Hub connection so the Blazor Server specific bits in here have to be updated to handle the 2fa login explicitly in a similar way to the normal sign in. I'll take a look.

DamianEdwards commented on src/BlazorIdentity/BlazorIdentity.csproj:
```diff
@@ -6,4 +6,8 @@
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
```
You can't reference that package in this project. The point of the BlazorIdentity project is to be usable in all Blazor projects and that package is only for use in ASP.NET Core projects, and even then, only in unsupported versions, as that package was deprecated a long time ago. You need to add types in this project that shim/wrap the actual Identity types, in this case I think that's `UserLoginInfo`.

grumpykiwi commented:
Ok thanks. Not sure how that got in there. Will take a look.

grumpykiwi committed:
Removed reliance on Identity pacjkage inside BlazorIdentity project. Added new UserLoginInfoResult class as a shim.

DamianEdwards commented on samples/BlazorIdentity.ServerApp/BlazorIdentity.ServerApp.csproj:
```diff
@@ -6,6 +6,10 @@
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="Data\AppUserLoginInfo.cs" />
```
This is strange, is this meant to be here?

DamianEdwards commented on src/BlazorIdentity.Server/Properties/launchSettings.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "profiles": {
```
This shouldn't have a `launchSettings.json` file as it's just a library.

DamianEdwards commented on src/BlazorIdentity/Properties/launchSettings.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Same as above, this shouldn't be here.

DamianEdwards commented on src/BlazorIdentity/UserLoginInfoResult.cs:
```diff
@@ -0,0 +1,13 @@
+using System.Text;
+using System.Threading.Tasks;
+
+public class UserLoginInfoResult
```
Let's just keep it the same name as the type it's shimming, like the other shim types in this library.

grumpykiwi committed:
Removed erroneous entry from project file.

grumpykiwi commented:
Ugh. I removed it from the project instead of deleting. Inadvertently added in the wrong place while solving for that UserLoginInfo dependency issue.

Fixed and pushed.


Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________
From: Damian Edwards ***@***.***>
Sent: Friday, July 15, 2022 12:33 PM
To: DamianEdwards/BlazorIdentity ***@***.***>
Cc: Mark Nash ***@***.***>; Author ***@***.***>
Subject: Re: [DamianEdwards/BlazorIdentity] Initial commit. Added support for register confirmation and personal … (PR #2)

[EXTERNAL SENDER]


@DamianEdwards requested changes on this pull request.

________________________________

In samples/BlazorIdentity.ServerApp/BlazorIdentity.ServerApp.csproj<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23discussion_r922381058&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=IU4zln5Xh%2F%2BtpFMKVdMA7aWtHgUbba5nqnCKWwHvaxU%3D&reserved=0>:

> @@ -6,6 +6,10 @@



     <ImplicitUsings>enable</ImplicitUsings>



   </PropertyGroup>







+  <ItemGroup>



+    <Compile Remove="Data\AppUserLoginInfo.cs" />



This is strange, is this meant to be here?

________________________________

In src/BlazorIdentity.Server/Properties/launchSettings.json<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23discussion_r922382783&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=zYKRJ0nffyBA9N%2BVuByLS5vVgdBOcClMtVRbtOAewHM%3D&reserved=0>:

> @@ -0,0 +1,12 @@



+{



+  "profiles": {



This shouldn't have a launchSettings.json file as it's just a library.

________________________________

In src/BlazorIdentity/Properties/launchSettings.json<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23discussion_r922383118&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=3w9qGxl%2Bo1uCysLjWH3sCdv96%2BYgZFI25%2By7xYNejjo%3D&reserved=0>:

> @@ -0,0 +1,12 @@



+{



Same as above, this shouldn't be here.

________________________________

In src/BlazorIdentity/UserLoginInfoResult.cs<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23discussion_r922383680&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=RS8lcOR84O4JYNEkGAoMSiKfs7tznL76h6Z8uKpMUok%3D&reserved=0>:

> @@ -0,0 +1,13 @@



+﻿namespace BlazorIdentity;



+using System;



+using System.Collections.Generic;



+using System.Linq;



+using System.Text;



+using System.Threading.Tasks;



+



+public class UserLoginInfoResult



Let's just keep it the same name as the type it's shimming, like the other shim types in this library.

—
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23pullrequestreview-1040608042&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=LJOEG9aLzQIbQuFaMEwLMH65SITJpH%2F8xqnjOKd%2FefQ%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKOLVXZQ2OD64OTZJCESVW3VUGOGLANCNFSM53JLCB2Q&data=05%7C01%7Cmnash%40partnercomm.net%7Ce86f0170547b4ddc02aa08da66881fe2%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637935032102899762%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4sf6n8skeGdlmfZTSXUOHqh6fDL6tDHv9dBTNFxob4k%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

grumpykiwi committed:
Added new functionalit and fixed a few issues. Refer readme for details. Any questions or issues please let me know and I will fix ASAP.

DamianEdwards commented on README.md:
```diff
@@ -24,9 +24,16 @@ Currently supports the following operations in Blazor Server apps:
+**NOTE**: Several of the above require configuring a 3rd party provider such as Twillio or SendGrid for full testing and use.
+
+**Also fixed:**
+ - A dependency on Microsoft.AspNetCore.Identity that snuck into the BlazorIdentity project.
```
These lines aren't needed as these changes only came in this PR :smile:

DamianEdwards commented on samples/BlazorIdentity.ServerApp/Pages/Account/Manage/Email.razor:
```diff
@@ -1,9 +1,12 @@
 
+<PageTitle>Manage email</PageTitle>
 
+<h1>Manage Email</h1>
```
```suggestion
<h1>Manage email</h1>
```

DamianEdwards commented on samples/BlazorIdentity.ServerApp/Pages/Account/Manage/Email.razor:
```diff
@@ -1,9 +1,12 @@
 
+<h1>Manage Email</h1>
 
+<BlazorIdentity.UI.Manage.Email TUser="AppUser" />
```
nit: Add a blank line below this

DamianEdwards commented on samples/BlazorIdentity.ServerApp/wwwroot/js/Alert.js:
```diff
@@ -0,0 +1,14 @@
+﻿function DismissAlert(divName) {
+    var mydiv = document.getElementById(divName);
+    mydiv.setAttribute("hidden", "true")
+    mydiv.hidden = "true";
+    document.getElementById(divName).hidden = true;
+}
+
+function ShowAlert(divName) {
+//    var mydiv = document.getElementById(divName).hidden = false;
+//    mydiv.removeAttribute("hidden");
+//    mydiv.hidden = "false";
+//    mydiv.setAttribute("hidden", "false")
+    document.getElementById(divName).hidden = false;
+}
```
There shouldn't be any custom JS in the server app for any of the UI components. If the UI needs JS then that should be packaged in the UI project and invoked via the JS runtime interop features of Blazor.

DamianEdwards replied:
There should be no custom JS in the sample app, any JS required to support the functionality should be in one of the library projects and invoked by the components or classes via JS interop.

As for alerting in general, I'd leave it up to the hosting page to manage that aspect of the UI and in the sample Blazor Server app I'm already using the BlazorStrap components to do that, e.g. https://github.com/DamianEdwards/BlazorIdentity/blob/afefa6a96012e5245b6f8957b60d7f658b7a4345/samples/BlazorIdentity.ServerApp/Pages/Account/Manage/Index.razor

DamianEdwards commented on src/BlazorIdentity.UI/Manage/DeletePersonalData.razor:
```diff
@@ -0,0 +1,112 @@
+        NavManager.NavigateTo("/");
+    }
+
+    protected override async Task OnInitializedAsync()
```
Put overrides before private handler methods.

DamianEdwards commented on src/BlazorIdentity.UI/ForgotPassword.razor:
```diff
@@ -0,0 +1,81 @@
+		if (user == null || !(await UserManager.IsEmailConfirmedAsync(user)))
+		{
+			// Don't reveal that the user does not exist or is not confirmed
+			NavManager.NavigateTo("/Account/ForgotPasswordConfirmation");
```
These components shouldn't navigate to pages that aren't in this library. For now, they should just raise events and let the hosting page handle that event and do the navigation.

DamianEdwards commented on src/BlazorIdentity.UI/BlazorIdentity.UI.csproj:
```diff
@@ -14,6 +14,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="7.0.0-preview.7.22330.6" />
     <PackageReference Include="Microsoft.AspNetCore.Components.Authorization" Version="7.0.0-preview.7.22330.6" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
```
This shouldn't be in here

DamianEdwards commented on src/BlazorIdentity.UI/BlazorIdentity.UI.csproj:
```diff
@@ -14,6 +14,8 @@
     <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="7.0.0-preview.7.22330.6" />
     <PackageReference Include="Microsoft.AspNetCore.Components.Authorization" Version="7.0.0-preview.7.22330.6" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
+    <PackageReference Include="Microsoft.AspNetCore.WebUtilities" Version="2.2.0" />
```
This shouldn't be in here either. What API do you need from it?

DamianEdwards commented on src/BlazorIdentity.UI/ConfirmEmail.razor:
```diff
@@ -0,0 +1,64 @@
+			StateHasChanged();
+			return;
+		}
+		ConfirmationCode = Encoding.UTF8.GetString(WebEncoders.Base64UrlDecode(ConfirmationCode));
```
Perhaps you want `System.Web.HttpUtility.UrlDecode`?

DamianEdwards commented on src/BlazorIdentity.UI/ConfirmEmailChange.razor:
```diff
@@ -0,0 +1,83 @@
+		StateHasChanged();
+	}
+
+	public async void ShowAlert()
+	{
+		await js.InvokeVoidAsync("ShowAlert", "myAlert");
+	}
+
+	public async void DismissAlert()
+	{
+		await js.InvokeVoidAsync("DismissAlert", "myAlert");
+	}
```
See notes above RE JS in the UI components themselves and the sample app already using BlazorStrap. Also, this component doesn't appear to have any other UI expect the alert, so what is displayed when the alert is dismissed?

DamianEdwards commented on src/BlazorIdentity.UI/Login.razor:
```diff
@@ -1,80 +1,112 @@
-    <div>
-        <button type="submit" class="btn btn-primary">Login</button>
-    </div>
+	<DataAnnotationsValidator />
+	<ValidationSummary />
+
+	<div class="mb-3">
+		<label for="email" class="form-label">Email address</label>
+		<InputText id="email" class="form-control" @bind-Value="loginModel.Email" />
+	</div>
+	<div class="mb-3">
+		<label for="password" class="form-label">Password</label>
+		<InputText id="password" type="password" class="form-control" @bind-Value="loginModel.Password" />
+	</div>
+	<div class="mb-3 form-check">
+		<InputCheckbox id="rememberMe" class="form-check-input" @bind-Value="loginModel.RememberMe" />
+		<label class="form-check-label" for="exampleCheck1">Remember me</label>
+	</div>
+	<div>
+		<button type="submit" class="btn btn-primary">Login</button>
+	</div>
+	<br />
+	<div>
+		<p>
+			<a href="/Account/ForgotPassword">Forgot your password?</a>
+		</p>
+		<p>
+			<a href="/Account/Register/@ReturnUrl">Register as a new user</a>
+		</p>
+		<p>
+			<a href="/Account/ResendEmailConfirmation">Resend email confirmation</a>
+		</p>
+	</div>
+
```
Seems this is all showing as changed due to you using tabs instead of spaces. Can you ensure all your changes use spaces please?

DamianEdwards commented on src/BlazorIdentity.UI/Login.razor:
```diff
@@ -1,80 +1,112 @@
+			return;
+		}
+
+		var result = await SignInManager.PasswordSignInAsync(loginModel.Email!, loginModel.Password!, loginModel.RememberMe, lockoutOnFailure: false);
+		if (result.Succeeded)
+		{
+			logger.LogWarning("User {user} logged in successfully", loginModel.Email);
+			await OnLogin.InvokeAsync();
+		}
+		if (result.RequiresTwoFactor)
+		{
+			NavManager.NavigateTo($"/Account/LoginWith2fa?RememberMe={loginModel.RememberMe}&ReturnUrl={ReturnUrl}");
+		}
+		if (result.IsLockedOut)
+		{
+			logger.LogWarning("User {Email} locked out.", loginModel.Email);
+			NavManager.NavigateTo("/Account/LockOut");
+		}
+		else
+		{
+			messageStore?.Add(() => loginModel, FailureMessage);
+			editContext.NotifyValidationStateChanged();
+		}
+	}
+
+	public class LoginModel
+	{
+		[Required]
+		[DataType(DataType.EmailAddress)]
+		[Display(Name = "Email address")]
+		public string? Email { get; set; }
+
+		[Required]
+		[DataType(DataType.Password)]
+		[Display(Name = "Password")]
+		public string? Password { get; set; }
+
+		[Display(Name = "Remember me?")]
+		public bool RememberMe { get; set; }
+	}
```
Tabs vs. Spaces again.

DamianEdwards commented on src/BlazorIdentity.UI/LoginWith2fa.razor:
```diff
@@ -0,0 +1,107 @@
+    </EditForm>
+<p>
+    Don't have access to your authenticator device? You can
+    <a id="recovery-code-login" href="/Account/LoginWithRecoveryCode">log in with a recovery code</a>.
```
I'm going to have think a bit more about this navigation model as having the component just blindly link to a URL that it doesn't itself carry a page for feels wrong. I don't think Blazor page components have the same capability as Razor Pages whereby a Razor library can bring in pages that are routable that can then be easily overridden by the hosting app so I don't think the same model will work here.

DamianEdwards reviewed (commented):
Please see comments in code.
There's a few open issues to work through regarding navigation and the general component architecture as we can't mimic what Identity UI does with Razor Pages today as far I know.

grumpykiwi commented:
Hi Damian,

I took a quick look at the comments and there several I can fix pretty quickly. There are a few others I will take another look at tomorrow.

I think it might not be a bad idea to discuss the big ones you pointed out:


 *   The alerts and subsequent JS. I was mimicking the Razor implementation on this, but agree there is a better way to do it. I could not for the life of me get the click to dismiss the alert to work without some JS. Since frontend is not my strong suit, I am probably doing something completely boneheaded.
 *   PersonalDataAttribute on DownloadPersonalData.razor. Maybe a shim of some sort around the AspNetCore.Identity implementation?
 *   Navigation to empty or non-existent pages
 *   Tabs vs spaces. Can this be set per project/solution?
 *   The existing issue with the 2FA cookie

I took a look at how the Razor pages identity implementation handles external logins. That is going to require some planning, Happy to help of course.

I can also be reached via my GitHub username at Hotmail. Feels a bit weird documenting my dirty laundry on GitHub &#128522;

Have a good evening

Cheers


Mark


Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________
From: Damian Edwards ***@***.***>
Sent: Thursday, July 21, 2022 7:37 PM
To: DamianEdwards/BlazorIdentity ***@***.***>
Cc: Mark Nash ***@***.***>; Author ***@***.***>
Subject: Re: [DamianEdwards/BlazorIdentity] Initial commit. Added support for register confirmation and personal … (PR #2)

[EXTERNAL SENDER]


@DamianEdwards commented on this pull request.

Please see comments in code.
There's a few open issues to work through regarding navigation and the general component architecture as we can't mimic what Identity UI does with Razor Pages today as far I know.

—
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FDamianEdwards%2FBlazorIdentity%2Fpull%2F2%23pullrequestreview-1047266063&data=05%7C01%7Cmnash%40partnercomm.net%7C3caa9c8c022b4c3bc3fd08da6b7a5040%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637940470328700864%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=qvVLpu3EHBJ9TCc7enmgPR%2B9Eyro0otDkIEFZ4VKzuc%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKOLVX4KJSZE5DJ7DNQOVQTVVHULLANCNFSM53JLCB2Q&data=05%7C01%7Cmnash%40partnercomm.net%7C3caa9c8c022b4c3bc3fd08da6b7a5040%7Cf2b31dc21677463d8cf12150d7ff1100%7C0%7C0%7C637940470328700864%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=iUoxBewPPfmnMct2D95MYSr4G25lRKXyvFPKYBtL%2B9Y%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

grumpykiwi committed:
Changes detailed in Blazor Identity Notes 07232022.md

grumpykiwi commented:
Hi

Just an FYI. I have started going thru the source code and documenting redirects in the UI project code, which components are called in the src/BlazorIdentity.UI project and callbacks in that project.

Once I have it documented I will fix any glaring inconsistencies. It hopefully should help with determining how redirects etc should be handled.

I should be able to finish the document off tomorrow some time.

Cheers


Mark

Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________

grumpykiwi committed:
Re-did personal data download based on this article:
https://docs.microsoft.com/en-us/aspnet/core/blazor/file-downloads?view=aspnetcore-6.0
Fixed a small re-direct bug in register

grumpykiwi committed:
Tweaked some returns. Added RegisterResult to allow multiple return values from callback. Added Excel file documenting calls and returns

grumpykiwi commented:
Good afternoon,

Sorry to bother you, but this issue is really starting to wear me out.

Quick question for you. The answer is probably quite simple. This may be due to my lack of knowledge on the deep plumbing of Identity etc. and probably OOP theory too.

I am fiddling around with implementing External Login. I have the call to Google and the subsequent callback firing correctly but having an issue at that point.

I am implementing the POST to external login with an API controller. Once I get the callback I attempt to do the following:

        var info = await _signInManager.GetExternalLoginInfoAsync();
        if (info == null)
        {
            ErrorMessage = $"Error loading external login information.";
            RedirectToPage("/account/login");
        }
        var userName = info!.Principal.Claims.Where(x => x.Type == ClaimTypes.Email).First().Value;
        var user = await _userManager.FindByEmailAsync(userName);

The call to FindByEmailAsync fails every time with an error saying it cannot find the SQL connection. This is not a problem with regular logins. Just external logins in this API controller. The constructor is like this:

public class AccountController : ControllerBase
{
    private readonly IBlazorSignInManager<AppUser> _signInManager;

    private readonly IBlazorUserManager<AppUser> _userManager;
    public string ErrorMessage { get; set; } = string.Empty;

    public AccountController(IBlazorSignInManager<AppUser> signInManager, IBlazorUserManager<AppUser> userManager)
    {
        _signInManager = signInManager;
        _userManager = userManager;
    }
}

Any ideas or thoughts on where I should look?

Cheers,


Mark

Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________

grumpykiwi committed:
Renamed shim for UserLoginInfo. Weird. Never got notified by GitHub about this one

grumpykiwi commented:
Good evening,

Go ahead and completely discard my previous email (shown below) about a weird issue with GetExternalLoginInfoAsync. I finally had some time to dig a little further and found that the code was looking at the OnConfigured property of the DBContext, which led me to the cause.  I had been fiddling with the new bits for SQL 2022 so changed the appsettings.json to point to a copy of the identity DB running that SQL instance inside a VM.  I neglected to change the settings back after discarding the VM. Hence, the timeouts.

Have a good weekend,


Mark


Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________
From: Mark Nash ***@***.***>
Sent: Monday, August 8, 2022 1:31 PM
To: DamianEdwards/BlazorIdentity ***@***.***>
Subject: Quick Question

Good afternoon,

Sorry to bother you, but this issue is really starting to wear me out.

Quick question for you. The answer is probably quite simple. This may be due to my lack of knowledge on the deep plumbing of Identity etc. and probably OOP theory too.

I am fiddling around with implementing External Login. I have the call to Google and the subsequent callback firing correctly but having an issue at that point.

I am implementing the POST to external login with an API controller. Once I get the callback I attempt to do the following:

        var info = await _signInManager.GetExternalLoginInfoAsync();
        if (info == null)
        {
            ErrorMessage = $"Error loading external login information.";
            RedirectToPage("/account/login");
        }
        var userName = info!.Principal.Claims.Where(x => x.Type == ClaimTypes.Email).First().Value;
        var user = await _userManager.FindByEmailAsync(userName);

The call to FindByEmailAsync fails every time with an error saying it cannot find the SQL connection. This is not a problem with regular logins. Just external logins in this API controller. The constructor is like this:

public class AccountController : ControllerBase
{
    private readonly IBlazorSignInManager<AppUser> _signInManager;

    private readonly IBlazorUserManager<AppUser> _userManager;
    public string ErrorMessage { get; set; } = string.Empty;

    public AccountController(IBlazorSignInManager<AppUser> signInManager, IBlazorUserManager<AppUser> userManager)
    {
        _signInManager = signInManager;
        _userManager = userManager;
    }
}

Any ideas or thoughts on where I should look?

Cheers,


Mark
Mark Nash
Senior Software Engineer

[cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]


PartnerComm, Inc.
Winner of 86 IABC Gold Quills
2022 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017

[cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]



Office: (817) 465-9277
***@***.******@***.***>
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________

grumpykiwi committed:
Added external login with exceptions noted in seperate document. I have looked at this off and on for several weeks trying to get around the issue but have have hit a wall. Decided to commit what I have.

Its probably awful and will end up on the cutting room floor. Sorry if that is the case. Best I could come up with right now. Happy to discuss.
